### PR TITLE
new software for simple keyframes selection

### PR DESCRIPTION
**Pull request by gregoire-dl**
- add first version of keyframeSelector software
- add FeedProvider support for reading RGB images
- CMake standardization of module requirements

### CODE REVIEW

File: src/cmakeFindModules/FindOpenImageIO.cmake [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#find headers
```
Comment (zvrba) [inline src/cmakeFindModules/FindOpenImageIO.cmake]: Why introduce yet another dependency instead of supporting only the formats that OpenCV supports?

Comment (simogasp) [inline src/cmakeFindModules/FindOpenImageIO.cmake]: I know, not enthusiastic either.
The thing is that OIIO can write exif while openCV can't. It's difficult to find a library for exif I/O with full support and the proper license.

Comment (zvrba) [inline src/cmakeFindModules/FindOpenImageIO.cmake]: How about this one: https://github.com/mayanklahiri/easyexif

Comment (simogasp) [inline src/cmakeFindModules/FindOpenImageIO.cmake]: it's already a dependency :-)
But it cannot write. Well we could add the writing part to that...

Comment (simogasp) [inline src/cmakeFindModules/FindOpenImageIO.cmake]: this one seems to have EXIF R/W too
http://freeimage.sourceforge.net/index.html

File: src/openMVG/dataio/ImageFeed.cpp [hunk 2]
```diff
@@ -28,15 +28,53 @@ class ImageFeed::FeederImpl
   
   FeederImpl(const std::string& imagePath, const std::string& calibPath);
   
-  bool readImage(image::Image<image::RGBColor> &imageRGB, 
+  template<typename T>
+  bool readImage(image::Image<T> &image,
```
Comment (simogasp) [inline src/openMVG/dataio/ImageFeed.cpp]: i don't know how it works, `readImage` is an abstract method of IFeed, does it need `override`?

File: src/openMVG/keyframe/KeyframeSelector.cpp [hunk 3]
```diff
@@ -91,208 +80,303 @@ void KeyframeSelector::process()
...
+    bool frameSelected = true;
+    auto& frameData = _framesData.at(frameIndex);
+    frameData.mediasData.resize(_feeds.size());
+
+    for(std::size_t mediaIndex = 0; mediaIndex < _feeds.size(); ++mediaIndex)
+    {
+      OPENMVG_COUT("media : " << _mediaPaths.at(mediaIndex) << std::endl);
+      auto& feed = *_feeds.at(mediaIndex);
+
+      if(frameSelected) // false if a camera of a rig is not selected
```
Comment (simogasp) [inline src/openMVG/keyframe/KeyframeSelector.cpp]:
```c++
if(!frameSelected)
{
  continue;
}
```

File: src/openMVG/keyframe/KeyframeSelector.hpp [hunk 4]
```diff
@@ -14,204 +14,310 @@
...
+   * @param[in] cameras informations
+   */
+  void setCameraInfos(const std::vector<CameraInfo>& CameraInfos)
+  {
+    _cameraInfos = CameraInfos;
+  }
+
+  /**
+   * @brief Set Sharpness selection preset
+   * @param sharpnessPreset enum
```
Comment (simogasp) [inline src/openMVG/keyframe/KeyframeSelector.hpp]: @param[in]

File: src/openMVG/keyframe/KeyframeSelector.hpp [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+   */
+  struct CameraInfo {
+    /// Camera brand
+    std::string brand = "Pinhole";
+    /// Camera model
+    std::string model = "Pinhole";
+    /// Focal length in mm or px
+    float focalLength = 1.2f;
+    /// If focalIsMM is false, focalLength is in px
+    bool focalIsMM = true;
```
Comment (gregoire-dl) [inline src/openMVG/keyframe/KeyframeSelector.hpp]: used in `KeyframeSelector.cpp` line 97

File: src/openMVG/keyframe/KeyframeSelector.hpp [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+
+  /**
+   * @brief Process media global informations
+   */
+  struct MediaInfo
+  {
+    /// height of the tile
+    unsigned int tileHeight = 0;
+    /// width of the tile
+    unsigned int tileWidth = 0;
```
Comment (gregoire-dl) [inline src/openMVG/keyframe/KeyframeSelector.hpp]: used in `KeyframeSelector.cpp` line 303

File: src/openMVG/keyframe/KeyframeSelector.hpp [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+
+  /**
+   * @brief Process media informations at a specific frame
+   */
+  struct MediaData
+  {
+    /// sharpness score
+    float sharpness = 0;
+    /// maximum distance score with keyframe media histograms
+    float distScore = 0;
```
Comment (gregoire-dl) [inline src/openMVG/keyframe/KeyframeSelector.hpp]: used in `KeyframeSelector.cpp` line 326

File: src/openMVG/keyframe/KeyframeSelector.hpp [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+
+  // Process structures
+
+  /**
+   * @brief Process media global informations
+   */
+  struct MediaInfo
+  {
+    /// height of the tile
+    unsigned int tileHeight = 0;
```
Comment (gregoire-dl) [inline src/openMVG/keyframe/KeyframeSelector.hpp]: used in `KeyframeSelector.cpp` line 302

File: src/openMVG/keyframe/KeyframeSelector.hpp [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+  /**
+   * @brief Camera informations
+   */
+  struct CameraInfo {
+    /// Camera brand
+    std::string brand = "Pinhole";
+    /// Camera model
+    std::string model = "Pinhole";
+    /// Focal length in mm or px
+    float focalLength = 1.2f;
```
Comment (gregoire-dl) [inline src/openMVG/keyframe/KeyframeSelector.hpp]: used in `KeyframeSelector.cpp` line 112

File: src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+file = open(args.input, "r")
+database = file.readlines()
+file.close()
+
+# process
+sensors = [entry.split(';')  for entry in database]
+sensors.sort()
+outDatabase = ""
+for entry in sensors:
+	outDatabase += ';'.join(entry)
```
Comment (simogasp) [inline src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py]: tab

File: src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
+import argparse
```
Comment (simogasp) [inline src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py]: `#!/usr/bin/python`

File: src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+# command line
+parser = argparse.ArgumentParser(description='Sort sensor width camera database')
```
Comment (simogasp) [inline src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py]: not really clear what it does... "Generate a copy of the sensor database sorted by increasing(?) sensor width"?

File: src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+parser.add_argument('-i', '--input', metavar='inputSensorDatabase.txt',required=True , type=str, 
+					help='input sensor width camera database')
```
Comment (simogasp) [inline src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py]: "File containing the original database of camera sensor widths"

File: src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+parser.add_argument('-o', '--output', metavar='outputSensorDatabase.txt', required=True , type=str,
+                    help='sorted output sensor width camera database')
```
Comment (simogasp) [inline src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py]: same here

File: src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py (hunk 12 above)
Comment (simogasp) [inline src/openMVG/exif/sensor_width_database/sensorDatabaseSort.py]: ah it just sort the database by the brand / model name, right? 🤦‍♂️

**Pull request merged**